src/controller/CommissioningWindowParams.h

```diff
@@ -0,0 +1,189 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCallback.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/NodeId.h>
+#include <lib/core/Optional.h>
+#include <lib/support/Span.h>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace Controller {
+
+// Passing SetupPayload by value on purpose, in case a consumer decides to reuse
+// this object from inside the callback.
+typedef void (*OnOpenCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status, SetupPayload payload);
+typedef void (*OnOpenCommissioningWindowWithVerifier)(void * context, NodeId deviceId, CHIP_ERROR status);
+typedef void (*OnOpenBasicCommissioningWindow)(void * context, NodeId deviceId, CHIP_ERROR status);
+
+template <typename Derived>
+class CommissioningWindowCommonParams
+{
+public:
+    CommissioningWindowCommonParams() = default;
+
+    NodeId GetNodeId() const
+    {
+        VerifyOrDie(mNodeId != kUndefinedNodeId);
+        return mNodeId;
+    }
+    // The node identifier of device
+    Derived & SetNodeId(NodeId nodeId)
+    {
+        mNodeId = nodeId;
+        return static_cast<Derived &>(*this);
+    }
+
+    System::Clock::Seconds16 GetTimeout() const { return mTimeout; }
+    // The duration for which the commissioning window should remain open.
+    Derived & SetTimeout(System::Clock::Seconds16 timeout)
+    {
+        mTimeout = timeout;
+        return static_cast<Derived &>(*this);
+    }
+    Derived & SetTimeout(uint16_t timeoutSeconds) { return SetTimeout(System::Clock::Seconds16(timeoutSeconds)); }
+
+    // The PAKE iteration count associated with the PAKE Passcode ID and
+    // ephemeral PAKE passcode verifier to be used for this commissioning.
+    uint32_t GetIteration() const { return mIteration; }
+    Derived & SetIteration(uint32_t iteration)
+    {
+        mIteration = iteration;
+        return static_cast<Derived &>(*this);
+    }
+
+    // The long discriminator for the DNS-SD advertisement.
+    uint16_t GetDiscriminator() const { return mDiscriminator.Value(); }
+    bool HasDiscriminator() const { return mDiscriminator.HasValue(); }
```

Comment:
Why do we need HasDiscriminator? Does not seem to be used anywhere....

Reply:
Added the use in https://github.com/project-chip/connectedhomeip/pull/34209/commits/19c0ffef9361764d6881cd9bc5de06317785dfbf, alongwith `HasNodeId()` - returning invalid args error

Reply:
Makes sense.